### PR TITLE
Initial prototype - cleanup & first steps forward

### PR DESCRIPTION
**Pull request by sysliquid**
With this pull request we finally start turning the prototype into the actual research. 

Provided that you agree the below I plan to then use this pull request to discuss what I'd like to do to move forward. This is all in the commit log but since this is the first one, I've written everything out here for convenience as well.

Note that I've moved the train,valid,test to a separate notebook that I %run in the data pipeline notebook.

First, I think I fixed two issues:
1. it looks there was an erroneous subtraction by -1. Probably just a copy and paste error.
2. I think you were doing fill forward on returns, but we only want that on prices. Double check that I wasn't missing something there

I also started changing a lot of those in-place edits to just simply using multiple data frames. We are not memory constrained yet with a data size this small so better to create lots of new columns and new data so we can easily compare and find issues etc.

I also changed the index to an actual datetime (so we can use datetime functions built into pandas). If you agree, I want to move away from the static shift operator to using actual periods (like 5m, 10m etc). That way we don't have to manually compensate for the previous close overlap and we can easily change return frequency.

And I right justified the date.   This would have no impact on the data panel, but I like it that way conceptually better. at 11am today we know the 10-11am return but we do not know the 11-12pm return. 

So please review the above and make sure I haven't missed anything. Once you have reviewed we will discuss how to move forward.

### CODE REVIEW

**Comment by xia-natsu** [comment]
Pipeline - agree with you. Data / ML better to be in separated in different notebooks.
Issues:

  

>   it looks there was an erroneous subtraction by -1. Probably just a copy and paste error.
Yes it was my error.
>     I think you were doing fill forward on returns, but we only want that on prices. Double check that I wasn't missing something there

Yes ffill price is the right operation here, was in a hurry and ffilling returns was not correct.

> I also started changing a lot of those in-place edits to just simply using multiple data frames. We are not memory constrained yet with a data size this small so better to create lots of new columns and new data so we can easily compare and find issues etc.

Not for now but we can have some naming rule to define these dataframe stages.


> I also changed the index to an actual datetime (so we can use datetime functions built into pandas). If you agree, I want to move away from the static shift operator to using actual periods (like 5m, 10m etc). That way we don't have to manually compensate for the previous close overlap and we can easily change return frequency.

Thanks. This is better and more readable. I tend to use non-period buckets as they may have dependency on Pandas version if you don't have server admin, but in our case it should not be problematic.


One suggestion - for any magic number / parameter variables / constants, like **tperiod_returns**, I prefer to move them to the head of notebook and capitalize them,
which is for our readability and convenience to change parameter, how do you think?

**Comment by sysliquid** [comment]
> One suggestion - for any magic number / parameter variables / constants, like tperiod_returns, I prefer to move them to the head of notebook and capitalize them,
which is for our readability and convenience to change parameter, how do you think?

Good idea. I've renamed tperiod_returns to kTPERIOD_RETURNS and put it under imports at the top of the TrainValidTestDataPrep notebook.

>  I tends to use non-period buckets as they may have dependency on Pandas version if you don't have server admin, but in our case it should not be problematic.

That makes sense to me, lets keep with the manual shift operator for now. We have regularly spaced bars and a full set for each ticker by design so manual shift should be no problem.

> ... some naming rule to define these dataframe stages

I'm open to suggestions, the only think I can think is to enforce the use of data_xyz type names for dataframes containing the raw, intermediate, or staged data. 

I'll push these now.

I am turning my attention to the PCA model which I'll start pushing to this branch next for your comments.

**Comment by sysliquid** [comment]
I'm out of time for today, but I've started working on the PCA model. This is still prototype, but trying to work out a little bit more example of what I'm wanting to test with these models. I'm in a real rush today and have some issues + need to cleanup/fixup my model testing but anyway take a look and let me know your comments. We can continue the discussion here and on mattermost.

**Comment by sysliquid** [comment]
Pushed again.

Got hung up exploring the internals of the sklearn PCA class. They have a 'score' function that I lifted and then kind of abused. See what you think of it. Would welcome any comments. In particular would like to know if we can do something similar for benchmarking the AE and VAE models.

Also found that the PCA class get_covariance() function only returns the common/cross sectional portion of the covar matrix. Took me a while to figure out why I couldn't match their covar! Anyway I like the way I do it better, but would like to know what you think. Seems like they follow that paper I linked and I have not yet had time to read.

Take a look at what I am doing with the validation data. I'm fitting the model to train then using validation to see where the PC's drop off as far as usefulness. There is more to do here but take a look. 

Also other random things that I'm kind of concerned about. There seems to be a decent amount of rounding error stability issues. I'm working on windows and my experience with python is that moving to linux solves a lot of those. See my comments for more.

**Comment by xia-natsu** [comment]
Regarding to the precision of PCA covar part, a quick workaround would be fix the random state of model. However, let me also check the paper to confirm whether it is a simple random seed triggered numpy float precision problem or not.

> #set precision to the 1bps level for var/covar
**RANDOM_SEED = 123**
var_atol = 1e-8
model_to_test = PCA(n_components=5, **random_state=RANDOM_SEED**)
model_to_test.fit(data_train)

**Comment by sysliquid** [comment]
I imagine that random seed is for automatic selection of factors or something because there shouldn’t be anything random about the PCA procedure itself. I know the sklearn PCA follows that paper so maybe lots of non-standard things under the hood? I’ll try to read that paper on the plane today.
